sgx/src/enclave/enclave.rs

```diff
@@ -141,41 +115,63 @@ impl Thread {
         Some(Self { enc, tcs, fnc })
     }
 
-    /// Issues `EENTER` instruction to the enclave.
+    /// Enter an enclave.
     ///
-    /// TODO add more comprehensive docs
+    /// This function enters an enclave using `Leaf` and provides the
+    /// specified argument to the enclave in the `rdx` register. On success,
+    /// the value of the `rdx` register at enclave exit is returned. There
+    /// are two failure cases. If the enclave performs an asynchronous exit
+    /// (AEX), the details of the exception are returned. Otherwise, an
+    /// unknown error has occurred.
     #[inline(always)]
     pub fn enter(
         &self,
-        rdi: usize,
-        rsi: usize,
-        rdx: usize,
-        leaf: Leaf,
-        r8: usize,
-        r9: usize,
-    ) -> Result<(), Option<ExceptionInfo>> {
+        how: Leaf,
+        arg: impl Into<Register<usize>>,
+    ) -> Result<Register<usize>, Option<ExceptionInfo>> {
         const FAULT: i32 = -libc::EFAULT;
         const EEXIT: i32 = 0;
 
         let mut exc = MaybeUninit::<ExceptionInfo>::uninit();
+        let mut rdx: usize = arg.into().into();
+        let rax: i32;
+
+        unsafe {
+            asm!(
+                "push rbp",
+                "mov  rbp, rsp",
```

Comment:
```suggestion
                "mov  rsp, rbp",
```
Is the function prologue not that the base pointer should take the current stack pointer's value?

Reply:
You're thinking AT&T syntax. This is Intel syntax. So: `mov DST, SRC`.

Reply:
Never mind, I see the inline asm operand syntax is backwards compared to AT&T.

Reply:
@npmccallum Yep, just saw that.